File: Codeforces/158B - Taxi.py
# ==================================================
# Problem  :   158B - Taxi
# Run time :   0.186 sec.
# Language :   Python 3
# ==================================================

def main():
	from sys import stdin, stdout

	n = int(stdin.readline())
	s = map(int, stdin.readline().split())

	cnt1, cnt2, cnt3, cnt4 = 0, 0, 0, 0

	for x in s:
		if x == 1: cnt1  += 1
		elif x == 2: cnt2  += 1
		elif x == 3: cnt3  += 1
		else: cnt4  += 1

	ans = cnt4

	ans += cnt2 // 2
	cnt2 = cnt2 % 2

	if cnt1 <= cnt3:
		ans += cnt1
		ans += cnt2
		ans += cnt3 - cnt1
	else:
		ans += cnt3
		cnt1 -= cnt3
		ans += cnt1 // 4
		cnt1 %= 4

		rem = cnt1 + cnt2 * 2
		if rem > 0:
			ans += 1 if rem <= 4 else 2

	stdout.write(str(ans) + '\n')

if __name__ == '__main__':
	main()

File: Codeforces/158B - Taxi.cpp
// ==================================================
// Problem  :   158B - Taxi
// Run time :   0.062 sec.
// Language :   C++11
// ==================================================

#include <cstdio>
using namespace std;

int main()
{
    int n;

    while(~scanf("%d", &n)) {
        int s, cnt[5] = {0};

        while(n--) {
            scanf("%d", &s);
            ++cnt[s];
        }

        int ans = cnt[4];

        ans += cnt[2] >> 1;         // cnt[2] / 2
        cnt[2] = cnt[2] & 1;

        if(cnt[1] <= cnt[3]) {
            ans += cnt[1];
            ans += cnt[2];
            ans += cnt[3] - cnt[1];
        }
        else {                      // cnt[1] > cnt[3]
            ans += cnt[3];
            cnt[1] -= cnt[3];
            ans += cnt[1] >> 2;     // cnt[1] / 4
            cnt[1] %= 4;

            if(cnt[1] + cnt[2]*2 > 0)
                ans += (cnt[1] + cnt[2]*2 <= 4)? 1 : 2;
        }

        printf("%d\n", ans);
    }

    return 0;
}
